OBJC_INTEROP.md

```diff
@@ -0,0 +1,174 @@
+# _Kotlin/Native_ interoperability with Swift/Objective-C
+
+This documents covers some details of Kotlin/Native interoperability with
+Swift/Objective-C.
+
+## Usage
+
+Kotlin/Native provides bidirectional interoperability with Objective-C.
+Objective-C frameworks and libraries can be used in Kotlin code if
+properly attached to the build (system frameworks are attached by default).
+See e.g. "Interop libraries" in
+[Gradle plugin documentation](GRADLE_PLUGIN.md#building-artifacts).
+Swift library can be used in Kotlin code if its API is exported to Objective-C
+with `@objc`. Pure Swift modules are not yet supported.
+
+Kotlin module can be used in Swift/Objective-C code if compiled into a
+[framework](GRADLE_PLUGIN.md#framework). See [calculator sample](samples/calculator)
+as an example.
+
+## Mappings
+
+The table below shows how Kotlin concepts are mapped to Swift/Objective-C and vice versa.
+
+| Kotlin | Swift | Objective-C | Notes |
+| ------ | ----- |------------ | ----- |
+| `class` | `class` | `@interface` | [note](#name-translation) |
+| `interface` | `protocol` | `@protocol` | |
+| `constructor` | Initializer | Initializer | |
```

Comment:
I'm not sure if constructors map to initializers in a natural way for Kotlin developers, so there probably should be some note on `initBy(...)` and invocation ordering.

Background:
I'm playing around with Kotlin/Native based iOS app and finding initializers a bit confusing (having no background in Obj-C / Swift).
I'm trying out code-only approach, so I've removed storyboards and invoking this in `AppDelegate`:
```kotlin
val mainViewController = ViewController()
println("mainViewController ready [${mainViewController.hashCode()}]")
```
In my `ViewController` I've added logging to constructors and `initWith...` methods as following:
```kotlin
@ExportObjCClass
class ViewController: UIViewController {

    init {
        println("ViewController [${hashCode()}] ctor: common init")
    }

    constructor(): super() {
        println("ViewController [${hashCode()}] ctor: no-arg")
    }
    constructor(aDecoder: NSCoder): super(aDecoder) {
        println("ViewController [${hashCode()}] ctor: NSCoder")
    }
    constructor(nibNameOrNil: String?, bundle: NSBundle?): super(nibNameOrNil, bundle) {
        println("ViewController [${hashCode()}] ctor: nibNameOrNil")
    }

    override fun init(): UIViewController? {
        println("ViewController [${hashCode()}]: init")
        return initBy(ViewController())
    }
    override fun initWithCoder(aDecoder: NSCoder): UIViewController? {
        println("ViewController [${hashCode()}]: initWithCoder")
        return initBy(ViewController(aDecoder))
    }
    override fun initWithNibName(nibNameOrNil: String?, bundle: NSBundle?): UIViewController {
        println("ViewController [${hashCode()}]: initWithNibName")
        return initBy(ViewController(nibNameOrNil, bundle))
    }
    // ...
}
```

Here's the resulting output:
```
ViewController [1807837959]: initWithNibName
ViewController [1807837959] ctor: common init
ViewController [1807837959] ctor: nibNameOrNil
ViewController [1807837959] ctor: common init
ViewController [1807837959] ctor: no-arg
mainViewController ready [1807837959]
```

So there are several points I'm finding odd, which ask for being covered by interop documentation:
* No-arg constructor leads to `initWithNibName`. This is probably due to [`init(nibName:bundle:)`](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621359-init) being [designated initializer](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Initialization.html#//apple_ref/doc/uid/TP40014097-CH18-ID217).
* `initWithNibName` is being called before any constructor code. Looks like that's caused by the nature of initializers in Objective-C, so they are treated a bit differently from other methods. I see that `initBy` is intrinsic and [has a special handling](https://github.com/JetBrains/kotlin-native/blob/77139d58f822aaef72498570371c7448426a695d/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt#L1686) when invoked, but it's not documented in any way, so a note here might be helpful. And actual `initWith...` methods appear as usual methods in generated `.kt` stubs, so nothing reveals their real nature.
* Two different constructors being called on one object, with two calls to `init { ... }` block. I guess Objective-C base class is able to invoke another constructor of derived class, and I'm not sure if such behavior can be achieved with Kotlin/JVM, so this also is a surprise point.

Please let me know if I'm doing something wrong or if something should be filled as an issue -- I'll try to provide as much details as possible.

Reply:
@r4zzz4k overriding of initializers is tricky and considered to be redesigned.
The problem with your example is that you call `super`-constructor that is based on non-designated initializer. The compiler should prohibit this, however it doesn't.

Reply:
I see, so this is temporary state of thins. And so even if I want to use no-arg constructor without working with nibs it should still invoke `super(null, null)` to reference `UIViewController(nibNameOrNil: String?, bundle: NSBundle?)`. Is that correct? If so, something like this should be great to have for newcomers like me not to fail this trap:
> Note: When implementing constructors of Kotlin classes which are extending Objective-C ones make sure you are calling superclass constructor corresponding to it's [*designated initializer*](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Initialization.html#//apple_ref/doc/uid/TP40014097-CH18-ID217) (i.e. [`initWithNibName:bundle:`](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621359-initwithnibname) for `UIViewController`).

Reply:
> if I want to use no-arg constructor without working with nibs it should still invoke `super(null, null)` to reference `UIViewController(nibNameOrNil: String?, bundle: NSBundle?)`. 

Yes. Swift also has this requirement.

We will either add such a warning to this doc or forbid this case in the compiler.

Reply:
Maybe some high level description of initBy and current state of things around initializers makes sense indeed.

Reply:
@olonho btw, I'm working on some improvements to Objective-C initializers support, which would make `initBy` (and also calling/overriding `init*` methods) obsolete. So I suggest waiting until this work is done and then completing the documentation about initializers.

Reply:
And here it is: #1512